### ISSUE TITLE
ESBMC passes "nil" object as parameter when targetting individual function

### ISSUE THREAD

**Issue by intrigus-lgtm** [esbmc/esbmc#1965]
This is a reduction of https://github.com/esbmc/esbmc/blob/123faaa0ecda77a1ee98242ced8d20a84a2f5ab7/regression/esbmc-cpp11/cpp/istream-api/test-istream.cc
When running this test with `--function f`
```cpp
#include <cassert>

struct b
{
  int a = 22;
};
void f(b c)
{
  assert(c.a == 22);
}

```

it fails, because ESBMC passes a "nil" object as the parameter to `f`.
**Shouldn't ESBMC pass a fully initialized object?**

<details>
<summary>IR</summary>

```
f (c:@F@f#$@S@b#):
        // 20 file cpptest.pre.cpp line 9 column 3 function f
        ASSERT c::0.a == 22 // assertion c.a == 22
        // 21 file cpptest.pre.cpp line 9 column 3 function f
        IF !(c::0.a == 22) THEN GOTO 1
        // 22 file cpptest.pre.cpp line 9 column 3 function f
        0;

        // 23 file cpptest.pre.cpp line 10 column 1 function f
     1: END_FUNCTION // f
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

__ESBMC_main (__ESBMC_main):
         [SNIP]
        // 31 no location
        FUNCTION_CALL:  pthread_start_main_hook()
        // 32 no location
        FUNCTION_CALL:  f(nil)  // <-----------------------
        // 33 no location
        FUNCTION_CALL:  atexit_handler()
         [SNIP]
```
</details>

**Comment by XLiZHI** [comment]
```
void f(b c)
{
  assert(c.a == 22);
}
```
It doesn't matter what the parameters are passed in, When we use `--function`, the argument to the function is nondetermined.

see: https://github.com/esbmc/esbmc/blob/123faaa0ecda77a1ee98242ced8d20a84a2f5ab7/scripts/README#L51-L55

**Comment by intrigus-lgtm** [comment]
Okay, then the test is just broken, I guess.

The problem is that the `std::istream` instance (of the original test)  is non-deterministic meaning that its vptr is not initialized.
This is currently not a problem, but with some changes I am currently working on, this fails (because we're trying to use a pointer that is normally set in the ctor).

Do you think it would make sense to warn/fail users about this?
I.e. "if parameter type has vptr directly or indirectly, warn/fail".

**Comment by XLiZHI** [comment]
> but with some changes I am currently working on
What changes did you make?

In my understanding, the flag `--function` is designed only for C verification and currently does not take into account constructors present in C++ structs. Perhaps, we should consider calling constructors for structs in C++, but that depends on the purpose. Why are you using this flag?

@fbrausse do you have any thoughts?

**Comment by intrigus-lgtm** [comment]
> > but with some changes I am currently working on
>
> What changes did you make?

i tried to fix #1866 by adding better support for virtual inheritance.
But the problem is pretty much unrelated to my changes

> In my understanding, the flag `--function` is designed only for C verification and currently does not take into account constructors present in C++ structs. Perhaps, we should consider calling constructors for structs in C++, but that depends on the purpose. Why are you using this flag?

I did not add the flag myself, but that's just what the test is using:
https://github.com/esbmc/esbmc/blob/123faaa0ecda77a1ee98242ced8d20a84a2f5ab7/regression/esbmc-cpp11/cpp/istream-api/test.desc#L3

**Comment by fbrausse** [comment]
Indeed, all parameter values to functions run via `--function` are supposed to be non-deterministic. In case of ESBMC's view on C, just not generating any constraints on the parameter symbol is fine, here: https://github.com/esbmc/esbmc/blob/123faaa0ecda77a1ee98242ced8d20a84a2f5ab7/src/goto-symex/symex_function.cpp#L77-L80 There basically are no invalid representations of C types.

But for C++ (and possibly other languages), in particular wrt. vptrs, the situation is different, I agree. Even if you were to `reinterpret_cast<b>(something)`, C++ doesn't allow to access or make copies of it if `something` doesn't actually refer to a valid object compatible with `b`. So if `b` contains a vptr, the current behaviour could be unexpected, from a user point of view.

@mikhailramalho @lucasccordeiro @XLiZHI Would you say that nondet values of C++ types should be initialized somehow? E.g., by calling one of its constructors?

Basically, the question is what reasonable semantics ESBMC could assign for `--function f` in
```c++
struct A {
  virtual ~A(); // has vptr
  // A() = delete; // can't construct
  // A(const A &) = delete; // can't copy
};

void f(A a) {
  // calls a.~A();
}
```
Here, A's vptr should be initialized to something sane, otherwise ESBMC will/should always give VERIFICATION FAILED at the end of f() making `--function f` useless for many C++ inputs.

When you uncomment the deletion of the constructors above, however, a contradiction arises: Unless I'm missing something, there basically is no way in C++ to create an instance of `A` nor to pass a copy of it. So `--function f` then wouldn't make sense in a pure C++ interpretation even if ESBMC did some "sane" initialization of the `a` parameter.

IMO generally disallowing `--function` for C++ would go too far, though. There could be a warning, or even an error when ESBMC detects non-POD arguments...
